테스트 환경 셋업 및 블로그 도메인 테스트 코드 추가

hobiJeong opened the pull request:
<!-- 이슈 넘버 url -->

### Issue Number

#212

<!-- 내용 (필수) -->

### Description

- 테스트 코드 작성 편의를 위한 라이브러리 추가
  - faker-js, fishery
- test 환경 셋업
- blog 도메인 Entity 및 관련 Entity Factory 생성
- blog 도메인 내의 controller 제외 테스트 코드 작성
- 유틸타입 추가
- 기존 블로그 업데이트 로직 수정
- 블로그 DomainError 추가
- Guard 메서드 에러 로직 수정

<!-- 관련된 이슈 링크 (선택) -->

### Related Issue Link

1. #211

hobiJeong committed:
setup/#212: test 환경 셋업

hobiJeong committed:
feat/#212: 테스트를 위한 라이브러리 설치 및 gitignore 추가

hobiJeong committed:
feat/#212: 블로그 업데이트 기능 개선 및 유효성 검사 추가

- 블로그 업데이트 메서드에서 필드 업데이트를 간소화하고, 유효성 검사 로직을 추가하여 블로그 이름, 설명, dDay 시작 날짜의 길이를 검증함.
- 새로운 블로그 업데이트 이벤트를 생성하여 변경 사항을 추적할 수 있도록 함.
- 블로그 컨트롤러에서 예외 처리 로직을 추가하여 블로그 멤버가 아닐 경우와 유효성 검사 실패 시 적절한 HTTP 예외를 발생시킴.

hobiJeong committed:
feat/#212: 블로그 조회 쿼리 핸들러에 BlogReadModel 추가 및 관련 수정

- BlogReadModel 클래스를 새로 생성하여 블로그 데이터를 구조화함.
- find-one-blog-by-user-id 쿼리 핸들러에서 반환 타입을 BlogReadModel로 변경하여 일관성 있는 데이터 반환.
- parse-query-by-colon-and-transform-to-object.transformer.ts에서 SingleProperty를 Exclusive로 변경하여 타입 안전성 강화.
- db.type.ts 파일을 추가하여 PrismaService와 TransactionalAdapterPrisma를 결합한 DB 타입 정의.

hobiJeong committed:
feat/#212: Jest 설정 및 테스트 모듈 추가

- jest.config.ts 파일에서 테스트 경로를 수정하고, setupFilesAfterEnv 및 preset 추가.
- TestPrismaModule 및 TestPrismaService를 새로 생성하여 Prisma와의 통합 테스트를 위한 모듈 및 서비스 구현.
- db.type.ts 파일을 추가하여 TestPrismaService와 TransactionalAdapterPrisma를 결합한 DB 타입 정의.
- global-setup.ts에서 DATABASE_URL에 connection_limit 추가하여 데이터베이스 연결 제한 설정.
- setup-tests.ts 파일을 추가하여 @Transactional 데코레이터를 no-op으로 모킹.

hobiJeong committed:
feat/#212: 블로그, 사용자, 사용자 연결 엔티티 팩토리 추가

- 블로그, 사용자, 사용자 연결 엔티티에 대한 팩토리 클래스를 새로 생성하여 테스트 데이터 생성을 용이하게 함.
- 각 팩토리에서 buildAndInsert 및 buildListAndInsert 메서드를 구현하여 데이터베이스에 엔티티를 삽입할 수 있도록 함.
- 엔티티 생성 시 faker를 사용하여 랜덤 데이터를 생성하고, 필요한 경우 사용자 정의 파라미터를 지원함.
- entity.base.ts에서 generateEntityId 함수를 추가하여 고유 ID 생성을 통합함.

hobiJeong committed:
feat/#212: 테스트를 위한 모의 객체 추가

- 블로그 도메인 서비스, 이벤트 발행기, 블로그 및 사용자 리포지토리에 대한 모의 객체를 새로 생성하여 테스트 용이성을 높임.
- RequestContextService의 메서드를 모킹하여 요청 컨텍스트를 관리하는 유틸리티 함수 추가.

hobiJeong committed:
test/#212: 블로그 생성 및 수정 관련 테스트 추가

- CreateBlogCommandHandler 및 PatchUpdateBlogCommandHandler에 대한 테스트를 새로 작성하여 블로그 생성 및 수정 로직의 유효성을 검증함.
- 블로그 생성 시 유저 존재 여부, 기존 블로그 존재 여부, 커넥션 상태에 따른 예외 처리 테스트 추가.
- 블로그 수정 시 블로그 멤버 여부 및 존재 여부에 따른 예외 처리 및 수정 로직 검증 테스트 추가.
- 블로그 엔티티의 유효성 검사 및 도메인 이벤트 발생 여부를 확인하는 테스트도 포함됨.

hobiJeong committed:
test/#212: 블로그 사용자 연결 해제 도메인 이벤트 핸들러 테스트 추가

- BlogUserConnectionDisconnectDomainEventHandler에 대한 테스트를 새로 작성하여 사용자 연결 해제 시 블로그 삭제 로직을 검증함.
- 블로그가 존재하지 않을 경우와 존재할 경우에 대한 두 가지 시나리오를 포함하여, 각각의 동작을 확인하는 테스트를 추가함.
- mockBlogRepository를 사용하여 블로그 조회 및 삭제 동작을 모킹함.

hobiJeong committed:
doc/#212: RequestContextService 모킹 함수에 주석 추가

- createMockRequestContextService 함수에 대한 설명 주석을 추가하여 테스트에서의 사용 목적을 명확히 함.

hobiJeong committed:
test/#212: 블로그 생성 및 수정 테스트 개선

- CreateBlogCommandHandler 및 PatchUpdateBlogCommandHandler의 테스트에서 블로그 생성 및 수정 시 리포지토리 호출 여부를 검증하는 로직을 추가함.
- 유저가 존재하지 않을 경우 블로그 생성 및 수정 시 리포지토리의 create 및 update 메서드가 호출되지 않음을 확인하는 테스트를 포함함.
- 블로그 생성 시 호출된 리포지토리의 create 메서드에 대한 인자 검증을 제거하여 테스트의 간결성을 높임.

hobiJeong committed:
refactor/#212: TestPrismaService에서 PrismaClient 상속 방식 변경 및 불필요한 에러 핸들러 제거

- TestPrismaService에서 PrismaClient를 직접 상속하도록 수정하여 코드 간결성을 높임.
- onModuleInit 메서드에서 불필요한 에러 핸들러 로직을 제거하여 가독성을 개선함.

rrgks6221 commented on package.json:
```diff
@@ -36,6 +36,7 @@
   "dependencies": {
     "@aws-sdk/client-cloudwatch-logs": "^3.726.1",
     "@aws-sdk/client-s3": "^3.735.0",
+    "@faker-js/faker": "^9.8.0",
```
dependency에 추가된 패키지들은 배포환경에 필요한 패키지들인가요??
pg도 프리즈마의 하위 디펜던시로 있을거같아서 되도록이면 dev로 들어가면 좋을거같습니다

rrgks6221 commented on src/features/blog/commands/create-blog/__spec__/create-blog.command-handler.spec.ts:
```diff
@@ -0,0 +1,195 @@
+    jest.clearAllMocks();
+  });
+
+  describe('블로그를 생성하면', () => {
+    describe('유저가 존재하지 않으면', () => {
+      it('INVALID_TOKEN 예외가 발생한다.', async () => {
```
given when then 으로 작성한다면 유저가 존재하지 않는 경우 블로그를 생성하면 예외가 발생한다 순서로 될거같습니다. 아래 케이스는 순서는 맞네요

https://brunch.co.kr/@springboot/292

hobiJeong replied:
순서가 그게 좀 더 자연스럽겠네요

rrgks6221 commented on src/features/blog/commands/create-blog/__spec__/create-blog.command-handler.spec.ts:
```diff
@@ -0,0 +1,195 @@
+        ).rejects.toThrow(HttpUnauthorizedException);
+      });
+
+      expect(mockBlogDomainService.create).not.toHaveBeenCalled();
+      expect(mockBlogRepository.create).not.toHaveBeenCalled();
```
테스트코드 작성 시 확인하는 방법이 크게 2가지가 있는데요(오래전에 봐서 정확히 기억은 안납니다). input에 따른 output을 검증, 내부 동작의 검증(메서드 호출 여부같은)
내부 동작의 검증같은 경우는 내부 구현에 따라 계속 변경돼야해서 중요한 부분이 아니면 개인적으로는 검증하지 않는것을 선호합니다

hobiJeong replied:
같은 의견으로 구체적인 내부 구현은 항상 변할 수 있다고 생각해 고민했었습니다 감사합니다

rrgks6221 commented on src/features/blog/commands/create-blog/__spec__/create-blog.command-handler.spec.ts:
```diff
@@ -0,0 +1,195 @@
+      expect(mockBlogRepository.create).not.toHaveBeenCalled();
+    });
+
+    describe('유저가 존재하면', () => {
+      describe('기존에 생성한 블로그가 없다면', () => {
+        it('블로그가 생성된다.', async () => {
+          const [requesterUser, requestedUser] = userFactory.buildList(2);
+          const userConnection = userConnectionFactory.build({
+            status: UserConnectionStatus.ACCEPTED,
+            requesterId: requesterUser.id,
+            requestedId: requestedUser.id,
+            requestedUser: requestedUser.hydrateProps,
+            requesterUser: requesterUser.hydrateProps,
+          });
+          const blog = blogFactory.build({
+            memberIds: [requesterUser.id, requestedUser.id],
+            members: [requesterUser.hydrateProps, requestedUser.hydrateProps],
+            createdBy: requesterUser.id,
+            connectionId: userConnection.id,
+          });
+          const blogProps = blog.getProps();
```
describe으로 뎁스를 나눈다면 beforeach나 beforeAll같은 문법을 활용하는게 좋을거같습니다.
가독성 측면, 동일한 조건의 연속된 테스트 코드 작성 면에서

hobiJeong replied:
repository는 모킹해서 매번 값이 달라질 필요가 없기 때문에 말씀 주신대로 beforeAll로 처리했습니다

rrgks6221 commented on src/features/blog/commands/create-blog/__spec__/create-blog.command-handler.spec.ts:
```diff
@@ -0,0 +1,195 @@
+    });
+
+    describe('기존에 생성한 블로그가 있다면', () => {
+      it(`${BlogAlreadyExistsError.name} 에러가 발생한다.`, async () => {
```
도메인 에러같은데 도메인 에러를 한국어로 표현하는게 읽기 더 좋을거같습니다

rrgks6221 commented on src/libs/core/prisma/__spec__/test-prisma.module.ts:
```diff
@@ -0,0 +1,8 @@
+  providers: [TestPrismaService],
+  exports: [TestPrismaService],
+})
+export class TestPrismaModule {}
```
테스트를 위한 프로바이더를 선언하여 사용하는것까지는 동의하는데 모듈을 생성하여 사용하는것은 모듈 시스템과 DI 잘 사용하지 못하고 있는거같습니다. injection token을 활용하여 프로바이더만 변경할수는 없던건가요?

hobiJeong replied:
현재 cls를 이용한 Transactional 데코레이터로 트랜잭션을 관리하고 있습니다.
그리고 plugin에선 모듈을 import하고 token을 통해 provider를 주입해줘야만 하는 구조입니다.

개인적으로 테스트에 사용되는 애들은 __spec__ 디렉터리 외부의 코드와 완전히 격리시키고 싶어서 기존 PrismaModule에 굳이 Provider로 등록하지 않고 새로운 모듈을 만들었습니다

rrgks6221 commented on test/mock/utils/mock.util.ts:
```diff
@@ -0,0 +1,44 @@
+/**
+ * @description RequestContextService를 호출하는 테스트에서 사용
+ */
+export const createMockRequestContextService = () => {
```
많은곳에서 사용된다면 jest의 setupFilesAfterEnv 을 활용하는것도 좋을거같습니다

hobiJeong replied:
일단은 repository에선 쓰이지 않기도 해서 추후에 고려해볼 것 같습니다

rrgks6221 reviewed (commented):
개인적으로 테스트코드의 js코드는 간결해야한다고 생각하는데요 (테스트 범위가 넓다면 다소 복잡할 수 있는데 논리가 복잡하면 안됨 테스트 범위가 넓은것도 유즈케이스나 서비스 명세가 잘못됐을 경우를 생각해봐야함) 전체적으로 테스트코드가 복잡하게 느껴졌습니다. 테스트코드 작성의 미숙함 때문인지 잘못 설계된 도메인 또는 잘못 구현된 구현체 때문인지는 모르겠네요

hobiJeong committed:
refactor/#212: @faker-js/faker 패키지를 devDependencies에 추가 및 package-lock.json에서 제거

- @faker-js/faker 패키지를 dependencies에서 devDependencies로 이동하여 개발 환경에서만 사용하도록 수정함.

hobiJeong committed:
test/#212: 블로그 생성 및 수정 테스트 리팩토링

- CreateBlogCommandHandler 및 PatchUpdateBlogCommandHandler의 테스트 구조를 개선하여 가독성을 높임.
- 유저 존재 여부 및 블로그 존재 여부에 따른 예외 처리 테스트를 명확히 하고, 블로그 생성 및 수정 시 리포지토리 호출 검증 로직을 추가함.
- 블로그 생성 및 수정 시의 유효성 검증 로직을 정리하여 테스트의 간결성을 유지함.

hobiJeong committed:
refactor/#212: package.json 및 package-lock.json 업데이트

- 테스트 환경 db 구축을 위한 라이브러리 devDependency로 재설치
